en: add doc for canary upgrade tidb operator (#1137)
pingcap/docs-tidb-operator: en/canary-upgrade-tidb-operator.md

---
title: Perform a Canary Upgrade on TiDB Operator
summary: Learn how to perform a canary upgrade on TiDB Operator in Kubernetes.
---

# Perform a Canary Upgrade on TiDB Operator

> This document describes how to perform a canary upgrade on TiDB Operator. By canary upgrades, you can prevent TiDB Operator upgrade from causing unexpected impact on all the TiDB clusters in Kubernetes. After you confirm the impact of TiDB Operator upgrade or the upgraded TiDB Operator works stably, you can normally upgrade TiDB Operator.

Reviewer: ```suggestion
This document describes how to perform a canary upgrade on TiDB Operator. Using canary upgrades, you can prevent normal TiDB Operator upgrade from causing an unexpected impact on all the TiDB clusters in Kubernetes. After you confirm the impact of TiDB Operator upgrade or that the upgraded TiDB Operator works stably, you can normally upgrade TiDB Operator.
```

> **Note:**
>
> - You can perform a canary upgrade only on `tidb-controller-manager` and `tidb-scheduler`.
> - Canary upgrade is supported since v1.1.10. The version of your current TiDB Operator should be >= v1.1.10.

## Related parameters

To support canary upgrade, some parameters are added to the `values.yaml` file in the `tidb-operator` chart. See [Related parameters](deploy-multiple-tidb-operator.md#related-parameters) for details.

## Canary upgrade process

1. Configure selector for the current TiDB Operator:

    Refer to [Upgrade TiDB Operator](upgrade-tidb-operator.md). Add the following configuration in the `values.yaml` file, and upgrade TiDB Operator:

    ```yaml
    controllerManager:
      selector:
      - version!=canary
    ```

>     If you already performed the step above, skip to Step 2.

Reviewer: ```suggestion
    If you have already performed the step above, skip to Step 2.
```

2. Deploy the canary TiDB Operator:

    Refer to [Deploy TiDB Operator](deploy-tidb-operator.md). Add the following configuration in the `values.yaml` file, and deploy the canary TiDB Operator **in a different namespace** (such as `tidb-admin-canary`):

    ```yaml
    controllerManager:
      selector:
      - version=canary
    appendReleaseSuffix: true
    #scheduler:
    #  create: false
    ```

    > **Note:**
    >
    > * It is recommended to deploy the new TiDB Operator in a separate namespace.
    > * Set `appendReleaseSuffix` to `true`.
    > * If you do not need to perform a canary upgrade on `tidb-scheduler`, configure `scheduler.create: false`.
    > * If you configure `scheduler.create: true`, a scheduler named `{{ .scheduler.schedulerName }}-{{.Release.Name}}` will be created. To use this scheduler, configure `spec.schedulerName` in the `TidbCluster` CR to the name of this scheduler.

> 3. To test the canary upgrade of `tidb-controller-manager`, set labels for some TiDB cluster by running the following command:

Reviewer: ```suggestion
3. To test the canary upgrade of `tidb-controller-manager`, set labels for a TiDB cluster by running the following command:
```
